<a href="https://colab.research.google.com/github/MarinaWolters/Coding-Tracker/blob/master/W1_Data_Acquisition_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Overview of Part I

As we start our journey into Big Data Analytics, the first thing we need to do is **get the data** in the form we need for analysis!  We'll start with an overview of how to acquire and *wrangle* data.

This notebook gives us a series of examples for:

* Acquiring data from files and remote sources
* Information extraction over HTML content
* A basic "vocabulary" of operators over tables (the relational algebra)

* "Data wrangling" or integration:
  * Cleaning and filtering data, using rules and based operations
  * Linking data across dataframes or relations
  * The need for approximate match and record linking

## The Motivating Question
To illustrate the principles, we focus on the question of how old company CEOs and founders are.  The question was in part motivated by the following New York Times article:

* Founders of Successful Tech Companies Are Mostly Middle-Aged: https://www.nytimes.com/2019/08/29/business/tech-start-up-founders-nest.html?searchResultPosition=2

So let's test this hypothesis!

In [1]:
# Let's install some libraries useful for processing web data
# For string similarity
!pip3 install py_stringsimjoin

# # lxml to parse xml tree
!pip3 install lxml

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
# Imports we'll use through the notebook, collected here for simplicity

# For parsing dates and being able to compare
import datetime

# For fetching remote data
import urllib
import urllib.request

# Pandas dataframes and operations
import pandas as pd

# Numpy matrix and array operations
import numpy as np

# Sqlite is a simplistic sql database
import sqlite3

# Approximate string matching, see 
import py_stringsimjoin as ssj
import py_stringmatching as sm

# Data visualization
import matplotlib

# Lecture Module 1: Acquiring Data

To test our hypothesis, we might want:

1. A list of companies (and, for futher details, perhaps their lines of business)
2. A list of company CEOs
3. Ages of the CEOs

We'll go through each of these using real data from the web.

### Reading Structured Data Sources

Let's start by looking up data about companies.  An example of this is at:

https://gist.githubusercontent.com/jvilledieu/c3afe5bc21da28880a30/raw/a344034b82a11433ba6f149afa47e57567d4a18f/Companies.csv

which has some nicely detailed information about companies, their categories, when they were founded, etc.  Let's load this (remote) CSV file into a dataframe.

In [3]:
data = urllib.request.urlopen(\
       'https://gist.github.com/jvilledieu/c3afe5bc21da28880a30/raw/a344034b82a11433ba6f149afa47e57567d4a18f/Companies.csv')

company_data_df = pd.read_csv(data)
    

In [4]:
# Let's write it to SQL, and read it back
conn = sqlite3.connect('local.db')

company_data_df.to_sql("companies", conn, if_exists="replace", index=False)

pd.read_sql_query('select * from companies', conn)

permalink  \
0                                   /organization/waywire   
1                         /organization/tv-communications   
2                           /organization/rock-your-paper   
3                          /organization/in-touch-network   
4                                   /organization/n-plusn   
...                                                   ...   
47753                                 /organization/zzish   
47754  /organization/zznode-science-and-technology-co-ltd   
47755                           /organization/zzzzapp-com   
47756                          /organization/a-list-games   
47757                                     /organization/x   

                                name                   homepage_url  \
0                           #waywire         http://www.waywire.com   
1                 &TV Communications          http://enjoyandtv.com   
2                  'Rock' Your Paper   http://www.rockyourpaper.org   
3                  (In)Touch Network  http://www.InTouchNetwork.com   
4                         +n (PlusN)               http://plusn.com   
...                              ...                            ...   
47753                          Zzish           http://www.zzish.com   
47754  ZZNode Science and Technology          http://www.zznode.com   
47755          Zzzzapp Wireless ltd.         http://www.zzzzapp.com   
47756                  [a]list games      http://www.alistgames.com   
47757                          [x+1]       http://www.xplusone.com/   

                                                                         category_list  \
0                                           |Entertainment|Politics|Social Media|News|   
1                                                                              |Games|   
2                                                               |Publishing|Education|   
3      |Electronics|Guides|Coffee|Restaurants|Music|iPhone|Apps|Mobile|iOS|E-Commerce|   
4                                                                           |Software|   
...                                                                                ...   
47753                    |Analytics|Gamification|Developer APIs|iOS|Android|Education|   
47754                                                            |Enterprise Software|   
47755                                    |Web Development|Advertising|Wireless|Mobile|   
47756                                                                          |Games|   
47757                                                            |Enterprise Software|   

                      market funding_total_usd     status country_code  \
0                      News         1 750 000    acquired          USA   
1                      Games        4 000 000   operating          USA   
2                 Publishing           40 000   operating          EST   
3                Electronics        1 500 000   operating          GBR   
4                  Software         1 200 000   operating          USA   
...                      ...               ...        ...          ...   
47753             Education           320 000   operating          GBR   
47754   Enterprise Software         1 587 301   operating          CHN   
47755       Web Development            97 398   operating          HRV   
47756                 Games         9 300 000   operating         None   
47757   Enterprise Software        45 000 000   operating          USA   

      state_code         region         city  funding_rounds  founded_at  \
0             NY  New York City     New York               1  01/06/2012   
1             CA    Los Angeles  Los Angeles               2        None   
2           None        Tallinn      Tallinn               1  26/10/2012   
3           None         London       London               1  01/04/2011   
4             NY  New York City     New York               2  01/01/2012   
...          ...            ...     

## Companies' CEOs: a Web Table

Now we need to figure out who the CEOs are for corporations.  One place to look is Wikipedia, which has an HTML table describing the CEOs.

https://en.wikipedia.org/wiki/List_of_chief_executive_officers#List_of_CEOs

Pandas actually makes it easy to read HTML tables...

In [5]:
# Now let's read an HTML table!
# opens the wiki page and finds List of CEOs table
company_ceos_df = pd.read_html('https://en.wikipedia.org/wiki/List_of_chief_executive_officers#List_of_CEOs')[0]
company_ceos_df

Company             Executive  \
0                    Accenture           Julie Sweet   
1           Aditya Birla Group  Kumar Mangalam Birla   
2                Adobe Systems      Shantanu Narayen   
3                       Agenus         Garo H. Armen   
4                       Airbus       Guillaume Faury   
..                         ...                   ...   
170            Williams-Sonoma        Laura J. Alber   
171                        WWE     Stephanie McMahon   
172                Yum! Brands            Greg Creed   
173               Zillow Group           Rich Barton   
174  Zoom Video Communications             Eric Yuan   

                                                 Title    Since  \
0                                               CEO[1]     2019   
1                                          Chairman[2]  1995[2]   
2                       Chairman, president and CEO[3]     2007   
3                            Founder, chairman, CEO[4]     1994   
4                                               CEO[5]     2012   
..                                                 ...      ...   
170                             President and CEO[154]     2010   
171  Chairwoman and Co-CEO (Alongside Nick Khan) [155]     2022   
172                                           CEO[156]     2015   
173                                           CEO[157]     2019   
174                   President, Chairman and CEO[158]     2011   

                                                                                          Notes  \
0                                                               Succeeded Pierre Nanterme, died   
1                                              Part of the Birla family business house in India   
2                                                                           Formerly with Apple   
3                                                Founder of the Children of Armenia Fund (COAF)   
4                                                                       Succeeded Louis Gallois   
..                                                                                          ...   
170                                                                   Replaced W. Howard Lester   
171           Chairwoman of the executive committeeChairwoman since 2022 CEO since July 22 2022   
172                                                                Previously CEO for Taco Bell   
173  Co-founder and previously was Zillow's CEO for nearly a decade. Succeeded Spencer Rascoff.   
174                                 Previously corporate Vice President of engineering at Cisco   

        Updated  
0    2019-01-31  
1    2018-10-01  
2    2018-10-01  
3    2018-10-01  
4    2017-11-14  
..          ...  
170  2017-11-11  
171  2017-11-11  
172  2017-11-11  
173  2018-12-10  
174  2020-10-12  

[175 rows x 6 columns]

## The Problem Gets Harder... Extracting Structured Fields

So far we have companies and CEOs.  But we don't have information on how old the CEOs are!

For a solution, we're going to go back to Wikipedia -- this time looking at the web pages for the CEOs!

This involves "crawling" the CEO pages, and "scraping" the relevant content.  In other words we have to do *information extraction*.

We'll start by constructing a list of CEO web pages, from the Company CEO dataframe above.  For this, we need to take the names and do a bit of tweaking, for example adding underscores instead of spaces.

In [7]:
crawl_list = []

for executive in company_ceos_df['Executive']:
  crawl_list.append('https://en.wikipedia.org/wiki/' + executive.replace(' ', '_'))
 
crawl_list

['https://en.wikipedia.org/wiki/Julie_Sweet',
 'https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla',
 'https://en.wikipedia.org/wiki/Shantanu_Narayen',
 'https://en.wikipedia.org/wiki/Garo_H._Armen',
 'https://en.wikipedia.org/wiki/Guillaume_Faury',
 'https://en.wikipedia.org/wiki/Daniel_Zhang',
 'https://en.wikipedia.org/wiki/Andy_Jassy',
 'https://en.wikipedia.org/wiki/Lisa_Su',
 'https://en.wikipedia.org/wiki/Stephen_Squeri',
 'https://en.wikipedia.org/wiki/Joseph_R._Swedish',
 'https://en.wikipedia.org/wiki/Tim_Cook',
 'https://en.wikipedia.org/wiki/Lakshmi_Niwas_Mittal',
 'https://en.wikipedia.org/wiki/John_Stankey',
 'https://en.wikipedia.org/wiki/Charles_Woodburn',
 'https://en.wikipedia.org/wiki/Tapan_Singhel',
 'https://en.wikipedia.org/wiki/Carlos_Torres_Vila',
 'https://en.wikipedia.org/wiki/Brian_Moynihan',
 'https://en.wikipedia.org/wiki/C.S._Venkatakrishnan',
 'https://en.wikipedia.org/wiki/Warren_Buffett',
 'https://en.wikipedia.org/wiki/Hubert_Joly',
 'https://en.wikipe

In [8]:
# Use urllib.urlopen to crawl all pages in crawl_list, and store the response of the page
# in list pages 

pages = []

for url in crawl_list:
    page = url.split("/")[-1]     # extracts the person name at the end of the url
    print('Looking at file %s' % page)

    # An issue: some of the accent characters won't work.  We need to convert them
    # into an HTML URL.  We'll split the URL, then use "parse.quote" to change
    # the structure, then re-form the URL
    url_list = list(urllib.parse.urlsplit(url))
    url_list[2] = urllib.parse.quote(url_list[2])
    url_ascii = urllib.parse.urlunsplit(url_list)
    try:
      response = urllib.request.urlopen((url_ascii))
      #Save page and url for later use.
      pages.append(response)
    except urllib.error.URLError as e:
      print(e.reason)


Looking at file Julie_Sweet
Looking at file Kumar_Mangalam_Birla
Looking at file Shantanu_Narayen
Looking at file Garo_H._Armen
Looking at file Guillaume_Faury
Looking at file Daniel_Zhang
Looking at file Andy_Jassy
Looking at file Lisa_Su
Looking at file Stephen_Squeri
Looking at file Joseph_R._Swedish
Looking at file Tim_Cook
Looking at file Lakshmi_Niwas_Mittal
Looking at file John_Stankey
Looking at file Charles_Woodburn
Looking at file Tapan_Singhel
Looking at file Carlos_Torres_Vila
Looking at file Brian_Moynihan
Looking at file C.S._Venkatakrishnan
Looking at file Warren_Buffett
Looking at file Hubert_Joly
Looking at file Sunil_Bharti_Mittal
Looking at file Stephen_A._Schwarzman
Looking at file Andrew_Mackenzie
Looking at file Oliver_Zipse
Looking at file Dave_Calhoun
Looking at file Rich_Lesser
Looking at file Bob_Dudley
Looking at file Denise_Morrison
Looking at file Mark_Shuttleworth
Looking at file Richard_Fairbank
Looking at file Jim_Umpleby
Looking at file Evan_Greenberg
L

In [9]:
len(pages)
pages

## Populating the Table with Executives

In [10]:
# Use lxml.etree.HTML(...) on the HTML content of each page to get a DOM tree that
# can be processed via XPath to extract the bday information.  Store the CEO name, 
# webpage, and the birthdate (born) in exec_df.

# We first check that the HTML content has a table of type `vcard`,
# and then extract the `bday` information.  If there is no birthdate, the datetime 
# value is NaT (not a type). 

from lxml import etree

exec_df = pd.DataFrame(columns=['name','page','born'])

for page in pages:    
    tree = etree.HTML(page.read().decode("utf-8"))    # creates a DOM tree of the page
    url = page.geturl()
    bday = tree.xpath('//table[contains(@class,"vcard")]//span[@class="bday"]/text()')
    if len(bday) > 0:
        name = url[url.rfind('/') +1:]                 # name is the part of the URL after the last /
        exec_df = exec_df.append({'name': name, 'page': url,
                   'born': datetime.datetime.strptime(bday[0], "%Y-%m-%d")}, ignore_index=True)   # 'born': datetime.datetime.strptime(bday[0], '%Y-%m-%d')
    else: 
            exec_df = exec_df.append({'name': url[url.rfind('/')+1:], 'page': url
                                      , 'born': None}, ignore_index=True)
        
exec_df

name                                                page  \
0             Julie_Sweet           https://en.wikipedia.org/wiki/Julie_Sweet   
1    Kumar_Mangalam_Birla  https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2        Shantanu_Narayen      https://en.wikipedia.org/wiki/Shantanu_Narayen   
3           Garo_H._Armen         https://en.wikipedia.org/wiki/Garo_H._Armen   
4         Guillaume_Faury       https://en.wikipedia.org/wiki/Guillaume_Faury   
..                    ...                                                 ...   
168        Laura_J._Alber        https://en.wikipedia.org/wiki/Laura_J._Alber   
169     Stephanie_McMahon     https://en.wikipedia.org/wiki/Stephanie_McMahon   
170            Greg_Creed            https://en.wikipedia.org/wiki/Greg_Creed   
171           Rich_Barton           https://en.wikipedia.org/wiki/Rich_Barton   
172             Eric_Yuan             https://en.wikipedia.org/wiki/Eric_Yuan   

                    born  
0                    NaT  
1    1967-06-14 00:00:00  
2    1963-05-27 00:00:00  
3    1953-01-31 00:00:00  
4    1968-02-22 00:00:00  
..                   ...  
168                 None  
169  1976-09-24 00:00:00  
170                 None  
171                 None  
172  1970-02-20 00:00:00  

[173 rows x 3 columns]

# Lecture Module 2: Data Transformation

Looking at our data to clean via *projection*...

Generally, we can extract one "narrower" table form another by using **double brackets**.

In [49]:
# Let's take a look at the data.  Here's a way of PROJECTING the exec_df dataframe into
# a smaller table

exec_df.index

RangeIndex(start=0, stop=173, step=1)

In [50]:
#exec_df['born'].index
#for i in range(len(exec_df)):
  #bday = exec_df['born'][i]
  #print(bday)
  #if bday == None or str(bday) == "NaT":
  #  continue
  #newbday = bday.strftime("%Y-%m-%d")
  #print(newbday)
  #exec_df['born'][i] = newbday

exec_df['born']

0                      NaT
1      1967-06-14 00:00:00
2      1963-05-27 00:00:00
3      1953-01-31 00:00:00
4      1968-02-22 00:00:00
              ...         
168                   None
169    1976-09-24 00:00:00
170                   None
171                   None
172    1970-02-20 00:00:00
Name: born, Length: 173, dtype: object

In [51]:
# df with 2 cols
exec_df[['name', 'born']]

name                 born
0             Julie_Sweet                  NaT
1    Kumar_Mangalam_Birla  1967-06-14 00:00:00
2        Shantanu_Narayen  1963-05-27 00:00:00
3           Garo_H._Armen  1953-01-31 00:00:00
4         Guillaume_Faury  1968-02-22 00:00:00
..                    ...                  ...
168        Laura_J._Alber                 None
169     Stephanie_McMahon  1976-09-24 00:00:00
170            Greg_Creed                 None
171           Rich_Barton                 None
172             Eric_Yuan  1970-02-20 00:00:00

[173 rows x 2 columns]

In [52]:
exec_df

name                                                page  \
0             Julie_Sweet           https://en.wikipedia.org/wiki/Julie_Sweet   
1    Kumar_Mangalam_Birla  https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2        Shantanu_Narayen      https://en.wikipedia.org/wiki/Shantanu_Narayen   
3           Garo_H._Armen         https://en.wikipedia.org/wiki/Garo_H._Armen   
4         Guillaume_Faury       https://en.wikipedia.org/wiki/Guillaume_Faury   
..                    ...                                                 ...   
168        Laura_J._Alber        https://en.wikipedia.org/wiki/Laura_J._Alber   
169     Stephanie_McMahon     https://en.wikipedia.org/wiki/Stephanie_McMahon   
170            Greg_Creed            https://en.wikipedia.org/wiki/Greg_Creed   
171           Rich_Barton           https://en.wikipedia.org/wiki/Rich_Barton   
172             Eric_Yuan             https://en.wikipedia.org/wiki/Eric_Yuan   

                    born            clean_name  
0                    NaT           Julie Sweet  
1    1967-06-14 00:00:00  Kumar Mangalam Birla  
2    1963-05-27 00:00:00      Shantanu Narayen  
3    1953-01-31 00:00:00         Garo H. Armen  
4    1968-02-22 00:00:00       Guillaume Faury  
..                   ...                   ...  
168                 None        Laura J. Alber  
169  1976-09-24 00:00:00     Stephanie McMahon  
170                 None            Greg Creed  
171                 None           Rich Barton  
172  1970-02-20 00:00:00             Eric Yuan  

[173 rows x 4 columns]

In [53]:
# If I use single brackets, I can extract a single column as a Series.
exec_df['name']

0               Julie_Sweet
1      Kumar_Mangalam_Birla
2          Shantanu_Narayen
3             Garo_H._Armen
4           Guillaume_Faury
               ...         
168          Laura_J._Alber
169       Stephanie_McMahon
170              Greg_Creed
171             Rich_Barton
172               Eric_Yuan
Name: name, Length: 173, dtype: object

In [54]:
# Notice anything awry?

for person in exec_df['name']:
    print (person)

Julie_Sweet
Kumar_Mangalam_Birla
Shantanu_Narayen
Garo_H._Armen
Guillaume_Faury
Daniel_Zhang
Andy_Jassy
Lisa_Su
Stephen_Squeri
Joseph_R._Swedish
Tim_Cook
Lakshmi_Niwas_Mittal
John_Stankey
Charles_Woodburn
Tapan_Singhel
Carlos_Torres_Vila
Brian_Moynihan
C.S._Venkatakrishnan
Warren_Buffett
Hubert_Joly
Sunil_Bharti_Mittal
Stephen_A._Schwarzman
Andrew_Mackenzie
Oliver_Zipse
Dave_Calhoun
Rich_Lesser
Bob_Dudley
Denise_Morrison
Mark_Shuttleworth
Richard_Fairbank
Jim_Umpleby
Evan_Greenberg
Chuck_Robbins
Jane_Fraser
James_Quincey
Brian_L._Roberts
Thomas_Gottstein
Ola_K%C3%A4llenius
Michael_Dell
Ed_Bastian
Christian_Sewing
Frank_Appel
Roland_Dickey_Jr.
Edward_D._Breen
G._V._Prasad
Devin_Wenig
Andrew_Wilson
B%C3%B6rje_Ekholm
Darren_Woods
Lisa_S._Jones
Carmine_Di_Sibio
Mark_Zuckerberg
Frederick_W._Smith
Rihanna
Sergio_Marchionne
Abigail_Johnson
James_Michael_Lafferty
James_Hackett
Terry_Gou
Lachlan_Murdoch
Daymond_John
Phebe_Novakovic
H._Lawrence_Culp_Jr.
Mary_T._Barra
Emma_Walmsley
David_M._Solom

In [55]:
# Removes underscores
def to_space(x):
  return x.replace('_', ' ')

# Let's use *apply* to call a function over each element, returning a new Series
exec_df['name'].apply(to_space)

0               Julie Sweet
1      Kumar Mangalam Birla
2          Shantanu Narayen
3             Garo H. Armen
4           Guillaume Faury
               ...         
168          Laura J. Alber
169       Stephanie McMahon
170              Greg Creed
171             Rich Barton
172               Eric Yuan
Name: name, Length: 173, dtype: object

In [56]:
# Let's use *apply* to call a function over each element, returning a new Series
exec_df['name'].apply(lambda x: x.replace('_', ' '))

0               Julie Sweet
1      Kumar Mangalam Birla
2          Shantanu Narayen
3             Garo H. Armen
4           Guillaume Faury
               ...         
168          Laura J. Alber
169       Stephanie McMahon
170              Greg Creed
171             Rich Barton
172               Eric Yuan
Name: name, Length: 173, dtype: object

In [57]:
# I can also use *apply* to call a function over the rows of a dataframe
exec_df.apply(lambda x: x['name'].replace('_', ' '), axis=1)

0               Julie Sweet
1      Kumar Mangalam Birla
2          Shantanu Narayen
3             Garo H. Armen
4           Guillaume Faury
               ...         
168          Laura J. Alber
169       Stephanie McMahon
170              Greg Creed
171             Rich Barton
172               Eric Yuan
Length: 173, dtype: object

In [58]:
# Let's clean the name by removing underscores and add it as a new col into our exec_df dataframe
exec_df['clean_name'] = exec_df['name'].apply(lambda x: x.replace('_', ' '))
exec_df

name                                                page  \
0             Julie_Sweet           https://en.wikipedia.org/wiki/Julie_Sweet   
1    Kumar_Mangalam_Birla  https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2        Shantanu_Narayen      https://en.wikipedia.org/wiki/Shantanu_Narayen   
3           Garo_H._Armen         https://en.wikipedia.org/wiki/Garo_H._Armen   
4         Guillaume_Faury       https://en.wikipedia.org/wiki/Guillaume_Faury   
..                    ...                                                 ...   
168        Laura_J._Alber        https://en.wikipedia.org/wiki/Laura_J._Alber   
169     Stephanie_McMahon     https://en.wikipedia.org/wiki/Stephanie_McMahon   
170            Greg_Creed            https://en.wikipedia.org/wiki/Greg_Creed   
171           Rich_Barton           https://en.wikipedia.org/wiki/Rich_Barton   
172             Eric_Yuan             https://en.wikipedia.org/wiki/Eric_Yuan   

                    born            clean_name  
0                    NaT           Julie Sweet  
1    1967-06-14 00:00:00  Kumar Mangalam Birla  
2    1963-05-27 00:00:00      Shantanu Narayen  
3    1953-01-31 00:00:00         Garo H. Armen  
4    1968-02-22 00:00:00       Guillaume Faury  
..                   ...                   ...  
168                 None        Laura J. Alber  
169  1976-09-24 00:00:00     Stephanie McMahon  
170                 None            Greg Creed  
171                 None           Rich Barton  
172  1970-02-20 00:00:00             Eric Yuan  

[173 rows x 4 columns]

In [59]:
# Renames the column 'name' with 'old_name'
exec_df.rename(columns={'name': 'old_name'})

old_name                                                page  \
0             Julie_Sweet           https://en.wikipedia.org/wiki/Julie_Sweet   
1    Kumar_Mangalam_Birla  https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2        Shantanu_Narayen      https://en.wikipedia.org/wiki/Shantanu_Narayen   
3           Garo_H._Armen         https://en.wikipedia.org/wiki/Garo_H._Armen   
4         Guillaume_Faury       https://en.wikipedia.org/wiki/Guillaume_Faury   
..                    ...                                                 ...   
168        Laura_J._Alber        https://en.wikipedia.org/wiki/Laura_J._Alber   
169     Stephanie_McMahon     https://en.wikipedia.org/wiki/Stephanie_McMahon   
170            Greg_Creed            https://en.wikipedia.org/wiki/Greg_Creed   
171           Rich_Barton           https://en.wikipedia.org/wiki/Rich_Barton   
172             Eric_Yuan             https://en.wikipedia.org/wiki/Eric_Yuan   

                    born            clean_name  
0                    NaT           Julie Sweet  
1    1967-06-14 00:00:00  Kumar Mangalam Birla  
2    1963-05-27 00:00:00      Shantanu Narayen  
3    1953-01-31 00:00:00         Garo H. Armen  
4    1968-02-22 00:00:00       Guillaume Faury  
..                   ...                   ...  
168                 None        Laura J. Alber  
169  1976-09-24 00:00:00     Stephanie McMahon  
170                 None            Greg Creed  
171                 None           Rich Barton  
172  1970-02-20 00:00:00             Eric Yuan  

[173 rows x 4 columns]

In [60]:
exec_df

name                                                page  \
0             Julie_Sweet           https://en.wikipedia.org/wiki/Julie_Sweet   
1    Kumar_Mangalam_Birla  https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2        Shantanu_Narayen      https://en.wikipedia.org/wiki/Shantanu_Narayen   
3           Garo_H._Armen         https://en.wikipedia.org/wiki/Garo_H._Armen   
4         Guillaume_Faury       https://en.wikipedia.org/wiki/Guillaume_Faury   
..                    ...                                                 ...   
168        Laura_J._Alber        https://en.wikipedia.org/wiki/Laura_J._Alber   
169     Stephanie_McMahon     https://en.wikipedia.org/wiki/Stephanie_McMahon   
170            Greg_Creed            https://en.wikipedia.org/wiki/Greg_Creed   
171           Rich_Barton           https://en.wikipedia.org/wiki/Rich_Barton   
172             Eric_Yuan             https://en.wikipedia.org/wiki/Eric_Yuan   

                    born            clean_name  
0                    NaT           Julie Sweet  
1    1967-06-14 00:00:00  Kumar Mangalam Birla  
2    1963-05-27 00:00:00      Shantanu Narayen  
3    1953-01-31 00:00:00         Garo H. Armen  
4    1968-02-22 00:00:00       Guillaume Faury  
..                   ...                   ...  
168                 None        Laura J. Alber  
169  1976-09-24 00:00:00     Stephanie McMahon  
170                 None            Greg Creed  
171                 None           Rich Barton  
172  1970-02-20 00:00:00             Eric Yuan  

[173 rows x 4 columns]

In [64]:
# We can do the same via SQL.  For the example we'll save the dataframe first...

# Let's convert types to string first to avoid errors.
temp_df = exec_df.applymap(str)

temp_df.to_sql('temp_exec', conn, if_exists="replace")
pd.read_sql_query('select name, replace(name, "_", " ") as clean_name from temp_exec', conn)

name            clean_name
0             Julie_Sweet           Julie Sweet
1    Kumar_Mangalam_Birla  Kumar Mangalam Birla
2        Shantanu_Narayen      Shantanu Narayen
3           Garo_H._Armen         Garo H. Armen
4         Guillaume_Faury       Guillaume Faury
..                    ...                   ...
168        Laura_J._Alber        Laura J. Alber
169     Stephanie_McMahon     Stephanie McMahon
170            Greg_Creed            Greg Creed
171           Rich_Barton           Rich Barton
172             Eric_Yuan             Eric Yuan

[173 rows x 2 columns]

## Selecting a subset of the rows

In [65]:
# Here's a column

exec_df['clean_name']

0               Julie Sweet
1      Kumar Mangalam Birla
2          Shantanu Narayen
3             Garo H. Armen
4           Guillaume Faury
               ...         
168          Laura J. Alber
169       Stephanie McMahon
170              Greg Creed
171             Rich Barton
172               Eric Yuan
Name: clean_name, Length: 173, dtype: object

In [70]:
# We can apply a test (predicate) to each column, returning a Series of boolean true/false values

exec_df['clean_name'] == 'Julie Sweet'

0       True
1      False
2      False
3      False
4      False
       ...  
168    False
169    False
170    False
171    False
172    False
Name: clean_name, Length: 173, dtype: bool

In [72]:
# If we compose these, we'll get only those rows where the boolean condition was True

exec_df[exec_df['clean_name'] == 'Julie Sweet']

name                                       page born   clean_name
0  Julie_Sweet  https://en.wikipedia.org/wiki/Julie_Sweet  NaT  Julie Sweet

In [73]:
# Returns row that's true + page col as a df  
# aka elem in df form 
exec_df[exec_df['clean_name'] == 'Julie Sweet'][['page']]

page
0  https://en.wikipedia.org/wiki/Julie_Sweet

In [27]:
# And in SQL...

temp_df= exec_df.applymap(str)
temp_df.to_sql('temp_exec', conn, if_exists="replace")

pd.read_sql_query('select * from temp_exec where clean_name="Julie Sweet"', conn)

index         name                                       page born  \
0      0  Julie_Sweet  https://en.wikipedia.org/wiki/Julie_Sweet  NaT   

    clean_name  
0  Julie Sweet

In [28]:
pd.read_sql_query('select page from temp_exec where clean_name="Julie Sweet"', conn)

page
0  https://en.wikipedia.org/wiki/Julie_Sweet

In [29]:
import numpy as np

exec_df.dropna(subset=['born'])

name                                                page  \
1    Kumar_Mangalam_Birla  https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2        Shantanu_Narayen      https://en.wikipedia.org/wiki/Shantanu_Narayen   
3           Garo_H._Armen         https://en.wikipedia.org/wiki/Garo_H._Armen   
4         Guillaume_Faury       https://en.wikipedia.org/wiki/Guillaume_Faury   
5            Daniel_Zhang          https://en.wikipedia.org/wiki/Daniel_Zhang   
..                    ...                                                 ...   
163         Doug_McMillon         https://en.wikipedia.org/wiki/Doug_McMillon   
165           Jason_Kilar           https://en.wikipedia.org/wiki/Jason_Kilar   
166        Charles_Scharf        https://en.wikipedia.org/wiki/Charles_Scharf   
169     Stephanie_McMahon     https://en.wikipedia.org/wiki/Stephanie_McMahon   
172             Eric_Yuan             https://en.wikipedia.org/wiki/Eric_Yuan   

                    born            clean_name  
1    1967-06-14 00:00:00  Kumar Mangalam Birla  
2    1963-05-27 00:00:00      Shantanu Narayen  
3    1953-01-31 00:00:00         Garo H. Armen  
4    1968-02-22 00:00:00       Guillaume Faury  
5    1972-01-11 00:00:00          Daniel Zhang  
..                   ...                   ...  
163  1966-10-17 00:00:00         Doug McMillon  
165  1971-04-26 00:00:00           Jason Kilar  
166  1965-04-24 00:00:00        Charles Scharf  
169  1976-09-24 00:00:00     Stephanie McMahon  
172  1970-02-20 00:00:00             Eric Yuan  

[106 rows x 4 columns]

In [30]:
pd.read_sql_query('select * from temp_exec where born is not null', conn).set_index('index')

name  \
index                         
0               Julie_Sweet   
1      Kumar_Mangalam_Birla   
2          Shantanu_Narayen   
3             Garo_H._Armen   
4           Guillaume_Faury   
...                     ...   
168          Laura_J._Alber   
169       Stephanie_McMahon   
170              Greg_Creed   
171             Rich_Barton   
172               Eric_Yuan   

                                                     page  \
index                                                       
0               https://en.wikipedia.org/wiki/Julie_Sweet   
1      https://en.wikipedia.org/wiki/Kumar_Mangalam_Birla   
2          https://en.wikipedia.org/wiki/Shantanu_Narayen   
3             https://en.wikipedia.org/wiki/Garo_H._Armen   
4           https://en.wikipedia.org/wiki/Guillaume_Faury   
...                                                   ...   
168          https://en.wikipedia.org/wiki/Laura_J._Alber   
169       https://en.wikipedia.org/wiki/Stephanie_McMahon   
170              https://en.wikipedia.org/wiki/Greg_Creed   
171             https://en.wikipedia.org/wiki/Rich_Barton   
172               https://en.wikipedia.org/wiki/Eric_Yuan   

                      born            clean_name  
index                                             
0                      NaT           Julie Sweet  
1      1967-06-14 00:00:00  Kumar Mangalam Birla  
2      1963-05-27 00:00:00      Shantanu Narayen  
3      1953-01-31 00:00:00         Garo H. Armen  
4      1968-02-22 00:00:00       Guillaume Faury  
...                    ...                   ...  
168                   None        Laura J. Alber  
169    1976-09-24 00:00:00     Stephanie McMahon  
170                   None            Greg Creed  
171                   None           Rich Barton  
172    1970-02-20 00:00:00             Eric Yuan  

[173 rows x 4 columns]

## Joining Data

We start with a simple join between company_ceos_df and exec_df and persist it to the database.  We then check how many companies did not have a match on CEO name.

In [31]:
exec_df[['clean_name', 'born']]

clean_name                 born
0             Julie Sweet                  NaT
1    Kumar Mangalam Birla  1967-06-14 00:00:00
2        Shantanu Narayen  1963-05-27 00:00:00
3           Garo H. Armen  1953-01-31 00:00:00
4         Guillaume Faury  1968-02-22 00:00:00
..                    ...                  ...
168        Laura J. Alber                 None
169     Stephanie McMahon  1976-09-24 00:00:00
170            Greg Creed                 None
171           Rich Barton                 None
172             Eric Yuan  1970-02-20 00:00:00

[173 rows x 2 columns]

In [32]:
company_ceos_df[['Executive', 'Company']]

Executive                    Company
0             Julie Sweet                  Accenture
1    Kumar Mangalam Birla         Aditya Birla Group
2        Shantanu Narayen              Adobe Systems
3           Garo H. Armen                     Agenus
4         Guillaume Faury                     Airbus
..                    ...                        ...
170        Laura J. Alber            Williams-Sonoma
171     Stephanie McMahon                        WWE
172            Greg Creed                Yum! Brands
173           Rich Barton               Zillow Group
174             Eric Yuan  Zoom Video Communications

[175 rows x 2 columns]

In [33]:
company_ceos_df[['Executive', 'Company']].merge(exec_df[['clean_name', 'born']], 
                                                left_on=['Executive'], 
                                                right_on=['clean_name'])

Executive                    Company            clean_name  \
0             Julie Sweet                  Accenture           Julie Sweet   
1    Kumar Mangalam Birla         Aditya Birla Group  Kumar Mangalam Birla   
2        Shantanu Narayen              Adobe Systems      Shantanu Narayen   
3           Garo H. Armen                     Agenus         Garo H. Armen   
4         Guillaume Faury                     Airbus       Guillaume Faury   
..                    ...                        ...                   ...   
167        Laura J. Alber            Williams-Sonoma        Laura J. Alber   
168     Stephanie McMahon                        WWE     Stephanie McMahon   
169            Greg Creed                Yum! Brands            Greg Creed   
170           Rich Barton               Zillow Group           Rich Barton   
171             Eric Yuan  Zoom Video Communications             Eric Yuan   

                    born  
0                    NaT  
1    1967-06-14 00:00:00  
2    1963-05-27 00:00:00  
3    1953-01-31 00:00:00  
4    1968-02-22 00:00:00  
..                   ...  
167                 None  
168  1976-09-24 00:00:00  
169                 None  
170                 None  
171  1970-02-20 00:00:00  

[172 rows x 4 columns]

In [34]:
temp_df= company_ceos_df.applymap(str)
temp_df.to_sql('company_ceos', conn, if_exists="replace")
temp_df= exec_df.applymap(str)
temp_df.to_sql('executives', conn, if_exists="replace")
company_data_df.to_sql('company_data', conn, if_exists='replace')

pd.read_sql_query('select Executive, Company, born from company_ceos ' +\
                  'join executives on Executive=clean_name ' +\
                  'where born is not null', conn)

Executive                    Company                 born
0             Julie Sweet                  Accenture                  NaT
1    Kumar Mangalam Birla         Aditya Birla Group  1967-06-14 00:00:00
2        Shantanu Narayen              Adobe Systems  1963-05-27 00:00:00
3           Garo H. Armen                     Agenus  1953-01-31 00:00:00
4         Guillaume Faury                     Airbus  1968-02-22 00:00:00
..                    ...                        ...                  ...
167        Laura J. Alber            Williams-Sonoma                 None
168     Stephanie McMahon                        WWE  1976-09-24 00:00:00
169            Greg Creed                Yum! Brands                 None
170           Rich Barton               Zillow Group                 None
171             Eric Yuan  Zoom Video Communications  1970-02-20 00:00:00

[172 rows x 3 columns]

###Finding the misses in the join.

Note that the join above resulted in 174 rows.  However, there are more rows in company_ceos_df so we are missing some companies.  We can see which are missed using a left outer join; setting "indicator= True" allows us to see which tuples in company_ceos_df failed to find a match (left_only, e.g. row 24 and 172).

In [35]:
pd.set_option('display.max_rows', 200)
display(company_ceos_df[['Executive', 'Company']].merge(exec_df[['clean_name', 'born']], 
                                                left_on=['Executive'], 
                                                right_on=['clean_name'], how="left", indicator=True))



Executive                          Company  \
0                       Julie Sweet                        Accenture   
1              Kumar Mangalam Birla               Aditya Birla Group   
2                  Shantanu Narayen                    Adobe Systems   
3                     Garo H. Armen                           Agenus   
4                   Guillaume Faury                           Airbus   
5                      Daniel Zhang                          Alibaba   
6                        Andy Jassy                           Amazon   
7                           Lisa Su           Advanced Micro Devices   
8                    Stephen Squeri                 American Express   
9                 Joseph R. Swedish                           Anthem   
10                         Tim Cook                            Apple   
11             Lakshmi Niwas Mittal                   Arcelor Mittal   
12                     John Stankey                             AT&T   
13                 Charles Woodburn                      BAE Systems   
14                    Tapan Singhel  Bajaj Allianz General Insurance   
15               Carlos Torres Vila  Banco Bilbao Vizcaya Argentaria   
16                   Brian Moynihan                  Bank of America   
17             C.S. Venkatakrishnan                         Barclays   
18                   Warren Buffett               Berkshire Hathaway   
19                      Hubert Joly                         Best Buy   
20              Sunil Bharti Mittal               Bharti Enterprises   
21            Stephen A. Schwarzman                 Blackstone Group   
22                 Andrew Mackenzie                              BHP   
23                     Oliver Zipse                              BMW   
24                     Dave Calhoun                           Boeing   
25                      Rich Lesser          Boston Consulting Group   
26                       Bob Dudley                               BP   
27                  Denise Morrison            Campbell Soup Company   
28                Mark Shuttleworth                        Canonical   
29                 Richard Fairbank            Capital One Financial   
30                      Jim Umpleby                      Caterpillar   
31                   Evan Greenberg                            Chubb   
32                    Chuck Robbins                    Cisco Systems   
33                      Jane Fraser                        Citigroup   
34                    James Quincey                        Coca-Cola   
35                 Brian L. Roberts                          Comcast   
36                 Thomas Gottstein                    Credit Suisse   
37                    Ola Källenius                       Daimler AG   
38                     Michael Dell                             Dell   
39                       Ed Bastian                  Delta Air Lines   
40                 Christian Sewing                    Deutsche Bank   
41                      Frank Appel                    Deutsche Post   
42                Roland Dickey Jr.            Dickey's Barbecue Pit   
43                  Edward D. Breen             Dow Chemical Company   
44                     G. V. Prasad                      Dr. Reddy's   
45                      Devin Wenig                             eBay   
46                    Andrew Wilson                  Electronic Arts   
47                     Börje Ekholm                         Ericsson   
48                     Darren Woods                       ExxonMobil   
49                    Lisa S. Jones                     EyeMail Inc.   
50                 Carmine Di Sibio                    Ernst & Young   
51                  Mark Zuckerberg                         Facebook   
52               Frederick W. Smith                            FedEx   
53                          Rihanna                     Fenty Beauty   
54                Sergio Marchionne                             Fi

In [36]:
pd.set_option('display.max_rows', 50)
result_df = company_ceos_df[['Executive', 'Company']].merge(exec_df[['clean_name', 'born']], 
                                                left_on=['Executive'], 
                                                right_on=['clean_name'], how="outer", indicator=True)

result_df[result_df['_merge'] != 'both']


Executive                  Company          clean_name  \
37     Ola Källenius               Daimler AG                 NaN   
47      Börje Ekholm                 Ericsson                 NaN   
88     Andrew Meslow  Bath & Body Works, Inc.                 NaN   
124  Michael O'Leary                  Ryanair                 NaN   
132     Roland Busch                  Siemens                 NaN   
177              NaN                      NaN  Ola K%C3%A4llenius   
178              NaN                      NaN   B%C3%B6rje Ekholm   
179              NaN                      NaN   Michael O%27Leary   

                    born      _merge  
37                   NaN   left_only  
47                   NaN   left_only  
88                   NaN   left_only  
124                  NaN   left_only  
132                  NaN   left_only  
177  1969-06-11 00:00:00  right_only  
178                  NaT  right_only  
179                 None  right_only

## Composing Joins



In [37]:
pd.read_sql_query('select Executive, Company, born  from company_ceos ' +\
                  'join executives on Executive=clean_name join company_data cd on Company=cd.name ' +\
                  'where born is not null', conn)

Executive                 Company                 born
0          Garo H. Armen                  Agenus  1953-01-31 00:00:00
1           Daniel Zhang                 Alibaba  1972-01-11 00:00:00
2             Andy Jassy                  Amazon  1968-01-13 00:00:00
3       Charles Woodburn             BAE Systems                  NaT
4      Mark Shuttleworth               Canonical  1973-09-18 00:00:00
5       Brian L. Roberts                 Comcast  1959-06-28 00:00:00
6            Devin Wenig                    eBay                  NaT
7        Mark Zuckerberg                Facebook  1984-05-14 00:00:00
8        Phebe Novakovic        General Dynamics                  NaT
9   H. Lawrence Culp Jr.        General Electric                  NaT
10         Sundar Pichai                  Google  1972-06-10 00:00:00
11       Darius Adamczyk               Honeywell  1966-02-08 00:00:00
12       Marillyn Hewson         Lockheed Martin  1953-12-27 00:00:00
13     Jonathan Ornstein          Mesa Air Group                 None
14  Robert James Thomson               News Corp  1961-03-11 00:00:00
15     Laxman Narasimhan               Starbucks  1967-05-15 00:00:00
16            Dave Lewis                   Tesco                 None
17             Elon Musk                 Twitter  1971-06-28 00:00:00
18             Elon Musk                 Twitter  1971-06-28 00:00:00
19         Hans Vestberg  Verizon Communications  1965-06-23 00:00:00

In [38]:
pd.read_sql_query('select * from company_ceos ' +\
                  'join executives on Executive=clean_name join company_data cd on Company=cd.name ' +\
                  'where born is not null', conn)

index                 Company             Executive  \
0       3                  Agenus         Garo H. Armen   
1       5                 Alibaba          Daniel Zhang   
2       6                  Amazon            Andy Jassy   
3      13             BAE Systems      Charles Woodburn   
4      28               Canonical     Mark Shuttleworth   
5      35                 Comcast      Brian L. Roberts   
6      45                    eBay           Devin Wenig   
7      51                Facebook       Mark Zuckerberg   
8      61        General Dynamics       Phebe Novakovic   
9      62        General Electric  H. Lawrence Culp Jr.   
10     66                  Google         Sundar Pichai   
11     72               Honeywell       Darius Adamczyk   
12     89         Lockheed Martin       Marillyn Hewson   
13     93          Mesa Air Group     Jonathan Ornstein   
14    102               News Corp  Robert James Thomson   
15    139               Starbucks     Laxman Narasimhan   
16    149                   Tesco            Dave Lewis   
17    154                 Twitter             Elon Musk   
18    154                 Twitter             Elon Musk   
19    158  Verizon Communications         Hans Vestberg   

                               Title Since  \
0          Founder, chairman, CEO[4]  1994   
1                             CEO[6]  2015   
2               CEO and president[7]  2021   
3        Chief Executive Officer[14]  2008   
4                Founder and CEO[28]  2017   
5               Chairman and CEO[35]  2002   
6              President and CEO[43]  2015   
7                   Chairman and CEO  2004   
8               Chairman and CEO[53]  2014   
9                   Chairman and CEO  2018   
10                           CEO[57]  2015   
11  Chairman, President, and CEO[63]  2017   
12   Chairman, president and CEO[77]  2013   
13              Chairman and CEO[81]  1998   
14                           CEO[89]  2013   
15            President and CEO[123]  2017   
16                    Group CEO[133]  2014   
17                          CEO[138]  2021   
18                          CEO[138]  2021   
19             Chairman and CEO[142]  2018   

                                                                                     Notes  \
0                                           Founder of the Children of Armenia Fund (COAF)   
1                                                                   Previously with Taobao   
2                                                                                      nan   
3                                                                       Succeeded Ian King   
4                                          Stepped down as CEO in 2009 and resumed in 2017   
5                                          Son of Ralph J. Roberts, the founder of Comcast   
6                             Third CEO of the company, after Meg Whitman and John Donahoe   
7                                                              Co-founded Facebook in 2004   
8   Worked for the Central Intelligence Agency and the United States Department of Defense   
9              Prior to joining GE, Culp worked at Danaher Corporation in Washington, D.C.   
10                                      Former director of Jive Software from 2011 to 2013   
11                                    Succeeded David M. Cote (now the executive chairman)   
12                     Also serves on the board of Sandia National Laboratories and DuPont   
13                                     Formerly led Virgin Express and Continental Express   
14                                   Previously managing editor of The Wall Street Journal   
15                            Previously CEO for Juniper Networks; succeeds Howard Schultz   
16                                                                Previously with Unilever   
17                                                                   Succeeded Jack Dorsey   
18         

In [39]:
company_ceos_df.merge(exec_df, 
                      left_on=['Executive'], 
                      right_on=['clean_name']).\
                      merge(company_data_df,
                            left_on='Company',
                            right_on='name')

Company             Executive  \
0                   Agenus         Garo H. Armen   
1                  Alibaba          Daniel Zhang   
2                   Amazon            Andy Jassy   
3              BAE Systems      Charles Woodburn   
4                Canonical     Mark Shuttleworth   
5                  Comcast      Brian L. Roberts   
6                     eBay           Devin Wenig   
7                 Facebook       Mark Zuckerberg   
8         General Dynamics       Phebe Novakovic   
9         General Electric  H. Lawrence Culp Jr.   
10                  Google         Sundar Pichai   
11               Honeywell       Darius Adamczyk   
12         Lockheed Martin       Marillyn Hewson   
13          Mesa Air Group     Jonathan Ornstein   
14               News Corp  Robert James Thomson   
15               Starbucks     Laxman Narasimhan   
16                   Tesco            Dave Lewis   
17                 Twitter             Elon Musk   
18                 Twitter             Elon Musk   
19  Verizon Communications         Hans Vestberg   

                               Title Since  \
0          Founder, chairman, CEO[4]  1994   
1                             CEO[6]  2015   
2               CEO and president[7]  2021   
3        Chief Executive Officer[14]  2008   
4                Founder and CEO[28]  2017   
5               Chairman and CEO[35]  2002   
6              President and CEO[43]  2015   
7                   Chairman and CEO  2004   
8               Chairman and CEO[53]  2014   
9                   Chairman and CEO  2018   
10                           CEO[57]  2015   
11  Chairman, President, and CEO[63]  2017   
12   Chairman, president and CEO[77]  2013   
13              Chairman and CEO[81]  1998   
14                           CEO[89]  2013   
15            President and CEO[123]  2017   
16                    Group CEO[133]  2014   
17                          CEO[138]  2021   
18                          CEO[138]  2021   
19             Chairman and CEO[142]  2018   

                                                                                     Notes  \
0                                           Founder of the Children of Armenia Fund (COAF)   
1                                                                   Previously with Taobao   
2                                                                                      NaN   
3                                                                       Succeeded Ian King   
4                                          Stepped down as CEO in 2009 and resumed in 2017   
5                                          Son of Ralph J. Roberts, the founder of Comcast   
6                             Third CEO of the company, after Meg Whitman and John Donahoe   
7                                                              Co-founded Facebook in 2004   
8   Worked for the Central Intelligence Agency and the United States Department of Defense   
9              Prior to joining GE, Culp worked at Danaher Corporation in Washington, D.C.   
10                                      Former director of Jive Software from 2011 to 2013   
11                                    Succeeded David M. Cote (now the executive chairman)   
12                     Also serves on the board of Sandia National Laboratories and DuPont   
13                                     Formerly led Virgin Express and Continental Express   
14                                   Previously managing editor of The Wall Street Journal   
15                            Previously CEO for Juniper Networks; succeeds Howard Schultz   
16                                                                Previously with Unilever   
17                                                                   Succeeded Jack Dorsey   
18                                                                   Succeeded Jack Dorsey   
19                                                                  Replaced Lo

# Lecture Module 3: Validating and Cleaning Data

How do we know our data is good?  We can create rules that trigger when the data fails some particular set of **constraints**.

In [40]:
# Test with validation rules

replace_item = ''

failed = False
for name in exec_df['clean_name']:
  if not name.replace(' ', replace_item).\
          replace('.', replace_item).\
          replace('\'',replace_item).\
          replace('-',replace_item).isalpha():
    print ("Illegal name %s"%name)
    failed = True
    
if failed:
  print('Found illegal names!')

Illegal name Ola K%C3%A4llenius
Illegal name B%C3%B6rje Ekholm
Illegal name Michael O%27Leary
Found illegal names!


## Data Cleaning: Fixing the Errors

One approach we could take is to realize that these are all strings in which accented characters are specially coded for use in (Wikipedia) URLs.  For instance, %C3 is a hex character code for an accented "u".  We can use a function called `unqote` to fix this...


In [41]:
from urllib.parse import unquote

exec_df['clean_name'].apply(unquote)

0               Julie Sweet
1      Kumar Mangalam Birla
2          Shantanu Narayen
3             Garo H. Armen
4           Guillaume Faury
               ...         
168          Laura J. Alber
169       Stephanie McMahon
170              Greg Creed
171             Rich Barton
172               Eric Yuan
Name: clean_name, Length: 173, dtype: object

## More Generally: Data Validation Tools

Are there tools to help us validate data?  Of course!  We see an [example](https://validators.readthedocs.io/en/latest/) of one such tool, simply called `validators`, below.  There are many others in the data cleaning literature.

In [42]:
!pip install validators

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for validators: filename=validators-0.20.0-py3-none-any.whl size=19581 sha256=ba25f10a5aa144b7f85110c1da47b16e8d8107e268ea04d251da65248dd203d5
  Stored in directory: /root/.cache/pip/wheels/19/09/72/3eb74d236bb48bd0f3c6c3c83e4e0c5bbfcbcad7c6c3539db8
Successfully built validators


In [43]:
import validators.url

# Are all of the URLs valid?
exec_df['page'].apply(validators.url)

0      True
1      True
2      True
3      True
4      True
       ... 
168    True
169    True
170    True
171    True
172    True
Name: page, Length: 173, dtype: bool

## Validation against a master list

We can also compare against "master" lists (in tables).  For example, company info about stock ticker symbols can be compared against the full list of symbols; states and countries can be compared against definitive lists.

Suppose we want to confirm that `company_data_df` has valid country codes.  


In [44]:
company_data_df[['name','country_code']]

name country_code
0                           #waywire          USA
1                 &TV Communications          USA
2                  'Rock' Your Paper          EST
3                  (In)Touch Network          GBR
4                         +n (PlusN)          USA
...                              ...          ...
47753                          Zzish          GBR
47754  ZZNode Science and Technology          CHN
47755          Zzzzapp Wireless ltd.          HRV
47756                  [a]list games          NaN
47757                          [x+1]          USA

[47758 rows x 2 columns]

TSome of these are NaNs -- let's drop them...

From someone's Github content, here's a nice list of all countries and region codes.

In [45]:
countries_df = pd.read_csv("https://raw.githubusercontent.com/lukes/ISO-3166-Countries-with-Regional-Codes/master/all/all.csv")

display(countries_df)

name alpha-2 alpha-3  country-code     iso_3166-2   region  \
0          Afghanistan      AF     AFG             4  ISO 3166-2:AF     Asia   
1        Åland Islands      AX     ALA           248  ISO 3166-2:AX   Europe   
2              Albania      AL     ALB             8  ISO 3166-2:AL   Europe   
3              Algeria      DZ     DZA            12  ISO 3166-2:DZ   Africa   
4       American Samoa      AS     ASM            16  ISO 3166-2:AS  Oceania   
..                 ...     ...     ...           ...            ...      ...   
244  Wallis and Futuna      WF     WLF           876  ISO 3166-2:WF  Oceania   
245     Western Sahara      EH     ESH           732  ISO 3166-2:EH   Africa   
246              Yemen      YE     YEM           887  ISO 3166-2:YE     Asia   
247             Zambia      ZM     ZMB           894  ISO 3166-2:ZM   Africa   
248           Zimbabwe      ZW     ZWE           716  ISO 3166-2:ZW   Africa   

             sub-region intermediate-region  region-code  sub-region-code  \
0         Southern Asia                 NaN        142.0             34.0   
1       Northern Europe                 NaN        150.0            154.0   
2       Southern Europe                 NaN        150.0             39.0   
3       Northern Africa                 NaN          2.0             15.0   
4             Polynesia                 NaN          9.0             61.0   
..                  ...                 ...          ...              ...   
244           Polynesia                 NaN          9.0             61.0   
245     Northern Africa                 NaN          2.0             15.0   
246        Western Asia                 NaN        142.0            145.0   
247  Sub-Saharan Africa      Eastern Africa          2.0            202.0   
248  Sub-Saharan Africa      Eastern Africa          2.0            202.0   

     intermediate-region-code  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  
..                        ...  
244                       NaN  
245                       NaN  
246                       NaN  
247                      14.0  
248                      14.0  

[249 rows x 11 columns]

In [46]:
validated = company_data_df.merge(countries_df, left_on=['country_code'], right_on=['alpha-3'],
                      how='left', indicator=True)

validated[validated['_merge'] != 'both']

permalink             name_x  \
8                /organization/0-6-com            0-6.com   
19          /organization/1000-markets       1000 Markets   
25              /organization/100du-tv           100du.tv   
30                 /organization/10sec              10sec   
35     /organization/115-network-disks  115 network disks   
...                                ...                ...   
47707               /organization/zura              Zura!   
47712            /organization/zursh-2              Zursh   
47729          /organization/zwoor-com          zwoor.com   
47743              /organization/zympi              Zympi   
47756       /organization/a-list-games      [a]list games   

                     homepage_url  \
8              http://www.0-6.com   
19     http://www.1000markets.com   
25           http://www.100du.com   
30                 http://10s.ec/   
35            http://www.115.com/   
...                           ...   
47707      http://www.zura.com.br   
47712       http://www.zursh.com/   
47729        http://www.zwoor.com   
47743        http://www.zympi.com   
47756   http://www.alistgames.com   

                                                              category_list  \
8                                                             |Curated Web|   
19                                            |Marketplaces|Art|E-Commerce|   
25                                                            |Hospitality|   
30                             |Social Commerce|E-Commerce|Mobile Commerce|   
35                                                              |Education|   
...                                                                     ...   
47707                                                          |E-Commerce|   
47712                                                                   NaN   
47729  |Meeting Software|Polling|Surveys|Android|iPhone|iPad|Events|Mobile|   
47743                                       |Networking|Web Hosting|Mobile|   
47756                                                               |Games|   

                  market funding_total_usd     status country_code state_code  \
8            Curated Web        2 000 000   operating          NaN        NaN   
19          Marketplaces          500 000    acquired          NaN        NaN   
25           Hospitality        3 000 000   operating          NaN        NaN   
30      Mobile Commerce         1 600 000         NaN          NaN        NaN   
35             Education              -     operating          NaN        NaN   
...                  ...               ...        ...          ...        ...   
47707        E-Commerce               -     operating          NaN        NaN   
47712                NaN           25 000   operating          NaN        NaN   
47729           Polling           100 000   operating          NaN        NaN   
47743       Web Hosting               -     operating          NaN        NaN   
47756             Games         9 300 000   operating          NaN        NaN   

      region_x  ... alpha-3  country-code iso_3166-2 region_y sub-region  \
8          NaN  ...     NaN           NaN        NaN      NaN        NaN   
19         NaN  ...     NaN           NaN        NaN      NaN        NaN   
25         NaN  ...     NaN           NaN        NaN      NaN        NaN   
30         NaN  ...     NaN           NaN        NaN      NaN        NaN   
35         NaN  ...     NaN           NaN        NaN      NaN        NaN   
...        ...  ...     ...           ...        ...      ...        ...   
47707      NaN  ...     NaN           NaN        NaN      NaN        NaN   
47712      NaN  ...     NaN           NaN        NaN      NaN        NaN   
47729      NaN  ...     NaN           NaN        NaN      NaN        NaN   
47743      NaN  ...     NaN           NaN        NaN      NaN        NaN   
47756      NaN  ...     NaN           NaN        NaN      NaN        NaN   

       intermedi

## Record Linking: Working around the Errors

Rather than figuring out how to clean these characters, we'll instead look at doing approximate matching.

Now we'll need to import some similarity matching code, to do approximate match between the original names and those returned by Wikipedia


In [47]:
# We are going to match the strings approximately, via "n-grams" or "q-grams" (sequences of n or q characters)
# Here it's five-grams

tok = sm.QgramTokenizer(qval=5,return_set=True)

In [48]:
# Now let's do a similarity join

# We'll reset the index, so there is a unique index field in the company_ceos_df dataframe
company_ceos_df.reset_index(inplace=True)

output_pairs = ssj.jaccard_join(company_ceos_df, exec_df, 'index', 'page', 'Executive', 'clean_name', tok, 0.35, 
                                l_out_attrs=['Executive'], r_out_attrs=['name'])

output_pairs[output_pairs['_sim_score'] < 1.0]

AssertionError: ignored

In [ ]:
# At last! Company info + CEO info, together!

total = company_ceos_df.merge(output_pairs,left_on=['Executive'],right_on=['l_Executive']).\
        merge(exec_df,left_on=['r_page'],right_on=['page']).\
        merge(company_data_df, left_on='Company', right_on='name', how="left")

total

In [ ]:
# Let's get ready to plot

%matplotlib inline

In [ ]:
total

## Simple Analysis of Linked Data: Grouping and Analytics

The `groupby` command allows us to coalesce data by groups.  In Pandas a Group is a special object with a set of rows.  We can see this with the `get_group` command.

In [ ]:
total[['born','Company','Executive']].sort_values('born')

In [ ]:
total.groupby(by='born').get_group(datetime.datetime.strptime('1945-08-24', '%Y-%m-%d'))[['Company','Executive','born']]

We can apply computations, such as `count`, to the items in the group.  Values that are empty (NaN) do not count (no pun intended).

In [ ]:
total.groupby(by='born').count()

In [ ]:
# We can do this in SQL too...

total.to_sql('total_info', conn, if_exists='replace', index=False)

# By default, SQL will include the Nan value (called NULL in SQL).  If we want to exclude it, 
# we need to include WHERE ... IS NOT NULL.

pd.read_sql('select born, count(Company) from total_info ' +\
            'where born is not null group by born', conn)

In [ ]:
# Let's look at when the CEOs were born

birthdays = total.groupby(by='born').count()[['Executive']]

birthdays

We can actually use `resample` over dates, with a parameter, to group eg by decade (10 years, where A is the code for year):

In [ ]:
birthdays.resample('10A').count().plot(kind='bar')

Maybe that's a little weird.  Can we do something more along the lines of what we expect, i.e., "1920s, 1930s, ..."?

In [ ]:
# Get rid of the nulls!
bdays = total[['born']].dropna()

bdays = bdays.applymap(lambda bday: str(int(bday.year / 10) * 10) + 's')

bdays.reset_index().groupby('born').count().plot(kind='bar')

# An Exercise

Is there a correlation between the kind of company and the age of the CEO?

Does the company's line of business matter?